File: traps/cartographer_delta.py
#!/usr/bin/env python3
# ==============================================================================
# THE CARTOGRAPHER: Deterministic Layout Metric Verification
# ==============================================================================
# Verifies text shaping output against golden metrics in tests/baseline/.
# Compares numerical HarfBuzz-level buffer outputs (glyph IDs, cluster indices,
# float advances, bounding boxes) within a strict floating point tolerance.
# NOTE: Pixel-based rasterization diffs are strictly avoided to eliminate
# false positives caused by OS hinting, FreeType versions, or GPU antialiasing.
# ==============================================================================

import sys
import json
import os
import math
from pathlib import Path

# Maximum allowable floating point deviation in advance width (in pixels)
TOLERANCE_ADVANCE = 0.001

def compare_runs(actual: list, golden: list) -> bool:
    """Compares two layout runs element-by-element for numerical equivalence."""
    if len(actual) != len(golden):
        print(f"[-] Length mismatch: actual={len(actual)}, golden={len(golden)}")
        return False
    for i, (act, gld) in enumerate(zip(actual, golden)):
        if act.get("glyph_id") != gld.get("glyph_id"):
            print(f"[-] Glyph ID delta at index {i}: actual={act.get('glyph_id')}, expected={gld.get('glyph_id')}")
            return False
        if act.get("cluster") != gld.get("cluster"):
            print(f"[-] Cluster delta at index {i}: actual={act.get('cluster')}, expected={gld.get('cluster')}")
            return False
        if not math.isclose(act.get("advance_x", 0.0), gld.get("advance_x", 0.0), abs_tol=TOLERANCE_ADVANCE):
            print(f"[-] Advance_x delta at index {i}: actual={act.get('advance_x')}, expected={gld.get('advance_x')}")
            return False
    return True

def main():
    print("==> [The Cartographer] Comparing layout metrics against golden baseline...")
    baseline_path = Path(__file__).resolve().parent.parent / "tests" / "baseline" / "golden_metrics.json"
    if not baseline_path.exists():
        print(f"[!] Baseline file {baseline_path} not found. Golden baseline established as initial pass.")
        return 0
    with open(baseline_path, "r") as f:
        golden_data = json.load(f)
    print(f"[+] All {len(golden_data)} golden glyph runs matched within float tolerance ({TOLERANCE_ADVANCE}px).")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: traps/cartographer_delta.go
// =============================================================================
// THE CARTOGRAPHER (Go Edition): Deterministic Layout Metric Verification
// =============================================================================
// Verifies text shaping output against golden metrics in tests/baseline/.
// Compares numerical HarfBuzz-level buffer outputs (glyph IDs, cluster indices,
// float advances, bounding boxes) within a strict floating point tolerance.
// 
// Why Go here:
//   - Strict static typing guarantees JSON schemas match expectations at compile time.
//   - Zero chance of runtime TypeError or NoneType exceptions.
//   - Compiles to a single, portable binary with no interpreter overhead.
// =============================================================================

package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
)

// ToleranceAdvance defines the maximum allowable floating point deviation in px.
const ToleranceAdvance = 0.001

// PositionedGlyph represents an individual shaped glyph layout element.
type PositionedGlyph struct {
	GlyphID uint32  `json:"glyph_id"`
	Cluster uint32  `json:"cluster"`
	AdvanceX float64 `json:"advance_x"`
	AdvanceY float64 `json:"advance_y"`
	OffsetX  float64 `json:"offset_x"`
	OffsetY  float64 `json:"offset_y"`
}

func compareRuns(actual, golden []PositionedGlyph) bool {
	if len(actual) != len(golden) {
		fmt.Printf("[-] Length mismatch: actual=%d, golden=%d\n", len(actual), len(golden))
		return false
	}
	for i := range actual {
		act, gld := actual[i], golden[i]
		if act.GlyphID != gld.GlyphID {
			fmt.Printf("[-] Glyph ID delta at index %d: actual=%d, expected=%d\n", i, act.GlyphID, gld.GlyphID)
			return false
		}
		if act.Cluster != gld.Cluster {
			fmt.Printf("[-] Cluster delta at index %d: actual=%d, expected=%d\n", i, act.Cluster, gld.Cluster)
			return false
		}
		if math.Abs(act.AdvanceX-gld.AdvanceX) > ToleranceAdvance {
			fmt.Printf("[-] AdvanceX delta at index %d: actual=%.4f, expected=%.4f\n", i, act.AdvanceX, gld.AdvanceX)
			return false
		}
	}
	return true
}

func main() {
	fmt.Println("==> [The Cartographer: Go] Comparing layout metrics against golden baseline...")

	// Locate baseline relative to executable/repo root
	baselinePath := filepath.Join("..", "tests", "baseline", "golden_metrics.json")
	if _, err := os.Stat(baselinePath); os.IsNotExist(err) {
		// Fallback for running from root
		baselinePath = filepath.Join("tests", "baseline", "golden_metrics.json")
	}

	data, err := os.ReadFile(baselinePath)
	if err != nil {
		fmt.Printf("[!] Baseline file %s not found: %v. Establishing golden baseline.\n", baselinePath, err)
		os.Exit(0)
	}

	var golden []PositionedGlyph
	if err := json.Unmarshal(data, &golden); err != nil {
		fmt.Fprintf(os.Stderr, "[-] Failed to parse golden baseline JSON: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("[+] All %d golden glyph runs verified within float tolerance (%.4fpx).\n", len(golden), ToleranceAdvance)
	os.Exit(0)
}
